[CustomOp] Add Dispatcher error callback

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #101015
* #101190
* #100980
* #100979

The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

zou3519 committed:
[CustomOp] Add Dispatcher error callback

The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

[ghstack-poisoned]

bdhirsh commented on torch/_custom_op.py:
```diff
@@ -786,3 +790,33 @@ def parse_return(annotation, error_fn):
+        raise NotImplementedError(
+            f"{custom_op}: when running with device='Meta' tensors: there is no "
+            f"abstract impl registered for this CustomOp. Please register one via "
+            f"CustomOp.impl_abstract to get this CustomOp to work with Meta tensors"
```
This will give a nice error if a user tries to run a custom op with meta tensors, and the op doesn't have a meta impl. Out of curiosity though, will we get an equally nice error for fake tensors? E.g. if you're using the standard `torch.compile()` workflow on a model with a custom op, and forgot an abstract impl. If not, it would be nice to add that too (which is probably easier since it's all in python).

zou3519 replied:
That's correct, we don't have a good error message for fake tensors yet, it is on the list of follow-ups

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -688,6 +688,16 @@ void initDispatchBindings(PyObject* module) {
+      [](c10::OperatorHandle& handle, py::object callback) {
+        auto obj = callback.ptr();
+        // SafePyObject steals a reference
+        Py_INCREF(obj);
```
It's more idiomatic to release() into SafePyObject, saves you an extra decref on exit

zou3519 commented:
PR is segfaulting in funny ways after rebase, may take a while to actually merge.

zou3519 committed:
Update on "[CustomOp] Add Dispatcher error callback"


The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[CustomOp] Add Dispatcher error callback"


The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[CustomOp] Add Dispatcher error callback"


The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[CustomOp] Add Dispatcher error callback"


The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[CustomOp] Add Dispatcher error callback"


The PyTorch Dispatcher's "no kernel found for DispatchKey" error message
is a bit long and winded. This PR adds a way to add a custom error
callback and changes the CustomOp API to use the custom error callback
to deliver better error messages.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Revert this as the earlier commits in the stack have been reverted' -c weird

zou3519 commented:
I'm going to submit a new stack
